src/core/lib/channel/channel_args.h

```diff
@@ -386,9 +395,10 @@ class ChannelArgs {
   }
   template <typename T>
   GRPC_MUST_USE_RESULT absl::enable_if_t<
-      std::is_same<
-          const grpc_arg_pointer_vtable*,
-          decltype(ChannelArgTypeTraits<std::shared_ptr<T>>::VTable())>::value,
+      std::is_same<const grpc_arg_pointer_vtable*,
```

Comment:
Not sure if this is better, but worth a try?

What happens if we write:

```
  GRPC_MUST_USE_RESULT absl::enable_if_t<
      std::is_same<
          const grpc_arg_pointer_vtable*,
          decltype(ChannelArgTypeTraits<std::shared_ptr<T>>::VTable())>::value,
      ChannelArgs>
  Set(absl::string_view name, std::shared_ptr<T> value) const {
     static_assert(SupportedSharedPtrType<T>::value);
```

that is - removing it from the enable_if, and hard failing the compilation if it isn't a supported type

we lose the ability to overload more supported types (seems we don't need that?)
we gain a pretty clear error message though?

Reply:
I'm not sure what you mean by losing the ability to overload more types, we can further specialize `SupportedSharedPtrType` right?

The error is much better, I like this approach.

```
In file included from test/core/channel/channel_args_test.cc:19:
./src/core/lib/channel/channel_args.h:399:5: error: static_assert failed due to requirement 'SupportedSharedPtrType<X>::value' "Type T must extend std::enable_shared_from_this to be added into ChannelArgs as a shared_ptr<T>"
```